File: src/main/scala/com/tomergabel/examples/patterns/PatternMatching0.scala
package com.tomergabel.examples.patterns

import scala.util.Try

/**
  * Created by tomerga on 04/04/2016.
  */
object PatternMatching0 extends App {

  case class HttpRequest(method: String,
                         path: String,
                         params: Map[String, Seq[String]] = Map.empty,
                         headers: Map[String, String] = Map.empty)

  case class HttpResponse(status: Int,
                          body: Array[Byte],
                          headers: Map[String, String] = Map.empty)

  def execute(req: HttpRequest): HttpResponse = {
    // Mock implementation, for example's sake
    val index = """<html><body>Hi!</body></html>"""
    val json = """{"response":"ok"}"""

    req match {
      case HttpRequest("GET", "/", _, _) =>
        HttpResponse(200, index.getBytes)

      case HttpRequest("GET", _, _, _) =>
        HttpResponse(404, Array.empty)

      case HttpRequest("POST", _, _, headers) if headers.get("Content-type").contains("application/json") =>
        HttpResponse(200, json.getBytes)

      case HttpRequest("POST", _, _, _) =>
        HttpResponse(415, Array.empty)

      case _ =>
        HttpResponse(405, Array.empty)
    }
  }

  def render(resp: HttpResponse) = s"[${resp.status}] ${new String(resp.body)}"
  println("Index: " + render(execute(HttpRequest("GET", "/"))))
  println("Invalid GET: " + render(execute(HttpRequest("GET", "/test"))))
  println("Valid POST: " + render(execute(HttpRequest("POST", "/", headers = Map("Content-type" -> "application/json")))))
  println("Invalid POST: " + render(execute(HttpRequest("POST", "/", headers = Map("Content-type" -> "application/xml")))))
  println("PUT: " + render(execute(HttpRequest("PUT", "/"))))
}

File: src/main/java/com/tomergabel/examples/java/PatternMatching0.java
package com.tomergabel.examples.java;

import java.util.Collections;
import java.util.Map;

import static java.util.Collections.emptyMap;

public class PatternMatching0 {

    private static class HttpRequest {
        private String method;
        private String path;
        private Map<String, String> headers;

        public String getMethod() { return method; }
        public String getPath() { return path; }
        public Map<String, String> getHeaders() { return headers; }

        public HttpRequest(String method, String path, Map<String, String> headers) {
            this.method = method;
            this.path = path;
            this.headers = headers;
        }
    }

    static class HttpResponse {
        private int status;
        private byte[] body;

        public int getStatus() { return status; }
        public byte[] getBody() { return body; }

        public HttpResponse(int status, byte[] body) {
            this.status = status;
            this.body = body;
        }
    }

    static private HttpResponse execute(HttpRequest req) {
        String index = "<html><body>Hi!</body></html>";
        String json = "{\"response\":\"ok\"}";
        byte[] empty = {};

        switch (req.getMethod()) {
            case "GET":
                if (req.getPath().equals("/"))
                    return new HttpResponse(200, index.getBytes());
                else
                    return new HttpResponse(404, empty);

            case "POST":
                if ("application/json".equals(req.getHeaders().get("Content-type")))
                    return new HttpResponse(200, json.getBytes());
                else
                    return new HttpResponse(415, empty);

            default:
                return new HttpResponse(405, empty);
        }
    }

    static private String render(HttpResponse resp) {
        return "[" + resp.getStatus() + "] " + new String(resp.getBody());
    }

    static public void main(String[] args) {
        Map<String, String> contentTypeHeader = Collections.singletonMap("Content-type", "application/json");

        System.out.println("Index: " + render(execute(new HttpRequest("GET", "/", emptyMap()))));
        System.out.println("Invalid GET: " + render(execute(new HttpRequest("GET", "/test", emptyMap()))));
        System.out.println("Valid POST: " + render(execute(new HttpRequest("POST", "/", contentTypeHeader))));
        System.out.println("Invalid POST: " + render(execute(new HttpRequest("POST", "/", emptyMap()))));
        System.out.println("PUT: " + render(execute(new HttpRequest("PUT", "/", emptyMap()))));
    }
}
